python/runtime/cudaq/algorithms/py_observe.cpp

```diff
@@ -416,6 +416,116 @@ provided for the specified number of circuit executions (`shots_count`).
 Returns:
   :class:`SampleResult`: A dictionary containing the measurement count results 
     for the :class:`Kernel`.)#");
+
+  mod.def(
+      "observe",
+      [&](py::object kernel,
+          std::variant<spin_op, std::vector<spin_op>> &spin_operator,
+          py::args args, int shots, std::optional<noise_model> noise,
+          std::size_t qpuId)
+          -> std::variant<observe_result, std::vector<observe_result>> {
+        // Do we have enough information to broadcast? And if so
+        // is the first arg a list
+        auto [disallowBroadCast, firstArgIsList] =
+            kernelCallableInputArgAnalysis(kernel);
+
+        // Get the kernel name
+        auto kernelName = py::hasattr(kernel, "kernelFunction")
+                              ? kernel.attr("kernelFunction")
+                                    .attr("__name__")
+                                    .cast<std::string>()
+                              : kernel.cast<kernel_builder<> &>().name();
+        auto &platform = cudaq::get_platform();
+        if (noise)
+          set_noise(*noise);
+
+        if (!disallowBroadCast && isBroadcastRequest(firstArgIsList, args)) {
+          auto argSet = createArgumentSet(args);
+          auto N = argSet.size();
+
+          std::vector<observe_result> results;
+          for (std::size_t currentIter = 0; auto &a : argSet) {
+            // Launch the observation task
+            auto ret = details::runObservation([&]() mutable { kernel(*a); },
+                                               std::get<0>(spin_operator),
+                                               platform, shots, kernelName, 0,
+                                               nullptr, currentIter, N)
+                           .value();
+            currentIter++;
+            results.push_back(ret);
+          }
+
+          platform.reset_noise();
+          return results;
+        }
+
+        // We are not broadcasting, regular observe call
+        if (qpuId >= platform.num_qpus())
+          throw std::runtime_error("Invalid qpu_id, platform only has " +
+                                   std::to_string(platform.num_qpus()) +
+                                   " qpus.");
+
+        // We could have been given a list of spin_ops, check here
+        spin_op op;
+        auto spinVariantIndex = spin_operator.index();
+        if (spinVariantIndex == 1) {
+          for (auto &o : std::get<std::vector<spin_op>>(spin_operator))
+            op += o;
+          op -= spin_op();
+        } else
+          op = std::get<spin_op>(spin_operator);
+
+        // Execute.
+        auto result =
+            details::runObservation([&]() mutable { kernel(*args); }, op,
+                                    platform, -1, kernelName, qpuId)
```

Comment:
Is there a reason for not sending on `shots` in this path?

Reply:
oversight